clientsv1_roles.go

```diff
@@ -0,0 +1,108 @@
+)
+
+// RolesServiceV1 provides client methods to interact with the
+// RolesService API v1.
+type RolesServiceV1 struct {
+	client *ClientV1
+}
+
+func (s *RolesServiceV1) newClient(ctx context.Context) clientsv1connect.RolesServiceClient {
+	return clientsv1connect.NewRolesServiceClient(
+		oauth2.NewClient(ctx, s.client.tokenSource),
+		s.client.gRPCURL(),
+		connect.WithInterceptors(s.client.defaultInterceptors...),
+	)
+}
+
+// RegisterResourcesMetadata is the metadata for a set of resources to be registered.
+type RegisterResourcesMetadata struct {
+	ResourceType roles.ResourceType
+	Revision     uuid.UUID
+}
+
+func (r RegisterResourcesMetadata) validate() error {
+	if !slices.Contains(roles.ResourceTypes(), r.ResourceType) {
+		return errors.New("invalid resource type")
+	}
+	return nil
+}
+
+// RegisterRoleResources registers the resources for a given resource type.
+// `resourcesIterator` is a function that returns a page of resources to register.
+// The function is invoked repeatedly until it produces an empty slice or an error.
+// If another replica is already registering the same resources, the function will return 0 without an error.
+//
+// Required scope: sams::roles.resources::write
+func (s *RolesServiceV1) RegisterRoleResources(ctx context.Context, metadata RegisterResourcesMetadata, resourcesIterator func() ([]*clientsv1.RoleResource, error)) (uint64, error) {
+	err := metadata.validate()
+	if err != nil {
+		return 0, errors.Wrap(err, "invalid metadata")
+	}
+
+	client := s.newClient(ctx)
+	stream := client.RegisterRoleResources(ctx)
+	// Metadata must be submitted first in the stream.
+	err = stream.Send(&clientsv1.RegisterRoleResourcesRequest{
+		Payload: &clientsv1.RegisterRoleResourcesRequest_Metadata{
+			Metadata: &clientsv1.RegisterRoleResourcesRequestMetadata{
+				ResourceType: string(metadata.ResourceType),
+				Revision:     metadata.Revision.String(),
+			},
+		},
+	})
+
+	sendResources := true
+	if err != nil {
+		// The stream has been closed; skip sending resources.
+		if errors.Is(err, io.EOF) {
+			sendResources = false
+		} else {
+			return 0, errors.Wrap(err, "failed to send metadata")
+		}
+	}
+	for sendResources {
+		resources, err := resourcesIterator()
+		if err != nil {
+			return 0, errors.Wrap(err, "failed to get resources")
+		}
+		if len(resources) == 0 {
+			break
+		}
+		err = stream.Send(&clientsv1.RegisterRoleResourcesRequest{
+			Payload: &clientsv1.RegisterRoleResourcesRequest_Resources_{
+				Resources: &clientsv1.RegisterRoleResourcesRequest_Resources{
+					Resources: resources,
+				},
+			},
+		})
+		if err != nil {
+			// The stream has been closed, so we stop sending resources.
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return 0, errors.Wrap(err, "failed to send resources")
+		}
+	}
+
+	resp, err := parseResponseAndError(stream.CloseAndReceive())
+	// Stream closed due to another replica registering the same resources.
+	if errors.Is(err, ErrAborted) {
```

Comment:
Is this something we do return ourselves in the RPC handler? It should be mentioned in the RPC docstring

Reply:
Also, do we interrupt the client mid-stream? If so, do we need to add handling on L98 (during the loop for `resourcesIterator()`)?

Reply:
On the server side I have a redislock that if a request fails to acquire (only tries once, does not wait) then we send an Aborted response.

The redislock is keyed based on the `resource_type` and the Actor's `ClientID`

Reply:
Since we abort if someone is already doing soemthing, should the error be one of the other codes, such as `resource_exhausted` (we're basically rate-limiting to 1) or `already_exists`?

https://connectrpc.com/docs/protocol#error-codes

Reply:
let's add some more in-depth docstrings here describing the behaviour in general

Reply:
I debated about which code would be best suited and didn't have a strong opinion on any of them so I am happy to switch it to something more suitable

> edit - ah, for ConnectRPC "aborted" does kind of have the meaning you are looking for:

yeah, still not entirely happy with it but it seemed at least somewhat reasonable.

> let's add some more in-depth docstrings here describing the behaviour in general

Do you want it on the RegisterRoleResources method added in this PR (which all consumers should be using to interface with the sdk) or on the underlying proto?

Reply:
- the RPC proto should document the aborted behaviour
- the SDK should document what it receive the aborted error
- the internals here should have some documentation on the behaviour we're expecting from the server

Reply:
I'm also okay with the current error code, but some docstring would go a long way here :) Maybe the translated SAMS-SDK error can include more details as well

Reply:
An alternative is to add a new field to the response e.g.
```
message RegisterRoleResourcesResponse {
  uint64 resource_count = 1;
  bool concurrency_aborted = 2;
}
```
Then we can send a normal response instead of using errors. wdyt?

Reply:
I think that wouldn't be good practice, documenting the error codes that get returned IMO is good